pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -57,975 +56,801 @@
+            raise MisconfigurationException(
+                f"You passed `Trainer(amp_type={self._amp_type_flag.value!r}, precision='bf16')` but "
+                "it's not supported. Try using `amp_type='native'` instead."
+            )
+        if self._precision_flag in (16, "bf16") and self._amp_type_flag == AMPType.APEX:
+            if isinstance(self.strategy, (DDPShardedStrategy, DDPSpawnShardedStrategy, DDPFullyShardedStrategy)):
+                raise MisconfigurationException(
+                    "Sharded plugins are not supported with apex, please switch to `amp_backend='native'`."
                 )
-            return self.distributed_backend
-
-        if self.use_gpu:
-            acc_cls = GPUAccelerator
-        elif self.use_tpu:
-            acc_cls = TPUAccelerator
-        elif self.use_ipu:
-            acc_cls = IPUAccelerator
-        else:
-            acc_cls = CPUAccelerator
-
-        accelerator = acc_cls()
-        return accelerator
-
-    def select_cluster_environment(self) -> ClusterEnvironment:
-        if self._cluster_environment is not None:
-            return self._cluster_environment
-        if self._is_slurm_managing_tasks():
-            rank_zero_info("Multiprocessing is handled by SLURM.")
-            return SLURMEnvironment()
 
-        for env_type in (BaguaEnvironment, TorchElasticEnvironment, KubeflowEnvironment, LSFEnvironment):
-            if env_type.detect():
-                return env_type()
-
-        return LightningEnvironment()
+    def _lazy_init_strategy(self) -> None:
+        """Lazily set missing attributes on the previously instantiated strategy."""
+        self.strategy.accelerator = self.accelerator
+        if self.precision_plugin:
+            self.strategy.precision_plugin = self.precision_plugin
+        if self.checkpoint_io:
+            self.strategy.checkpoint_io = self.checkpoint_io
+        if hasattr(self.strategy, "cluster_environment"):
+            self.strategy.cluster_environment = self.cluster_environment
+        if hasattr(self.strategy, "parallel_devices"):
+            if self.strategy.parallel_devices:
+                self._parallel_devices = self.strategy.parallel_devices
+            else:
+                self.strategy.parallel_devices = self._parallel_devices
+        if hasattr(self.strategy, "num_nodes"):
+            self.strategy._num_nodes = self._num_nodes_flag
+        if hasattr(self.strategy, "sync_batchnorm"):
+            self.strategy.sync_batchnorm = self.sync_batchnorm
+        if hasattr(self.strategy, "set_world_ranks"):
+            self.strategy.set_world_ranks()
 
-    def set_distributed_mode(self, strategy: Optional[str] = None):
+        from pytorch_lightning.utilities import _IS_INTERACTIVE
 
-        if strategy is None and self.is_training_type_in_plugins:
-            return
+        interactive_compatible_strategy = ("dp", "ddp_spawn", "ddp_sharded_spawn", "tpu_spawn")
```

Comment:
Let's use the Strategy.strategy_name instead, this could avoid possible bugs.

Reply:
updated